File: jain_machine_tools/jain_machine_tools/report/freight_charges/freight_charges.py


import frappe
from frappe import _
from frappe.utils import nowdate

def execute(filters=None):
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data


def get_columns(filters):
	return [
		{"label": _("Freight Charges"), "fieldname": "freight_charges", "fieldtype": "Link", "options": "Freight Charges", "width": 150},
		{"label": _("Sales Invoice"), "fieldname": "sales_invoice", "fieldtype": "Link", "options": "Sales Invoice", "width": 180},
		{"label": _("Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
		{"label": _("Supplier"), "fieldname": "supplier_name", "fieldtype": "Link", "options": "Supplier", "width": 300},
		{"label": _("Type"), "fieldname": "type", "fieldtype": "Data", "width": 200},
		{"label": _("Amount"), "fieldname": "amount", "fieldtype": "Data", "width": 170},
	] 
	
def get_data(filters):
	if not filters:
		filters = {}  
	
	data= []
	conditions = ""
	
	if filters.get("sales_invoice"):
		conditions += " AND si.name = %(sales_invoice)s"
	
	if filters.get("supplier"):
		conditions += " AND fcd.supplier_name = %(supplier)s"

	if filters.get("from_date") and filters.get("to_date"):
		conditions += " AND si.posting_date BETWEEN %(from_date)s AND %(to_date)s"
  
	sales_invoice = frappe.db.sql(f""" 
					SELECT 
						fc.name as freight_charges,
						fcd.sales_invoice as sales_invoice,
						si.posting_date ,
						fcd.supplier_name,
						fcd.type,
						fcd.amount
					FROM `tabFreight Charges` fc
					LEFT JOIN `tabFreight Charges Details` fcd ON fcd.parent = fc.name
					LEFT JOIN `tabSales Invoice` si ON fcd.sales_invoice = si.name
					WHERE 1=1
					{conditions}
					""",filters, as_dict=True)
	
	last_fc = None
	for row in sales_invoice:
		if row.freight_charges != last_fc:
			data.append({
				'indent': 0,
				'freight_charges': row.freight_charges,
			})
			last_fc = row.freight_charges
   
		data.append({
				'indent': 1,
				'freight_charges' : '',
				'sales_invoice' : row.sales_invoice,
				'posting_date' : row.posting_date,
				'supplier_name' : row.supplier_name,
				'type' : row.type,
				'amount' : row.amount,
			}
		)
	
	return data
	

File: jain_machine_tools/jain_machine_tools/report/freight_charges/freight_charges.js
// Copyright (c) 2026, Praxon Technovation and contributors
// For license information, please see license.txt

frappe.query_reports["Freight Charges"] = {
	"filters": [
		{
            fieldname: "sales_invoice",
            label: __("Sales Invoice"),
            fieldtype: "Link",
            options: "Sales Invoice",
        },
		{
            fieldname: "supplier",
            label: __("Supplier"),
            fieldtype: "Link",
            options: "Supplier",
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
        },
	],
	formatter(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (!data) return value;

        if (data.indent === 0) {
            value = `<b>${value}</b>`;
        }
        
        return value;
    },
};
